# Start and installation

Import the module to your python environment. You will need to install the packages first by running `pip install -r requirements.txt` at the root of this document.

# First steps

Create a file called `cred.txt` on the same folder from where you'll be using the script. This file has to contain (In order):
```
USER=<your_aap_username>
PASSWORD=<your_aap_password>
ROOT=<DSP_api_root>
```

The API root can either be `https://submission-test.ebi.ac.uk/api/` for the -test environments or `https://submission.ebi.ac.uk/api/` for the production environment.

Please check the root of the repository if you have doubts about what the user and the password are.

`cred.txt` is ignored by the repository, so don't fear to accidentally upload it if you push changes.

# Running the script

Once everything has been set up, we can begin with the code:

In [1]:
import DSP_submission as ds # Import the object

dsp = ds.DspCLI()

The code is documented and everything can be accessed through the `help()` command in python. For example, if we would like to know all the methods available to the user and a description:

In [2]:
help(dsp)

Help on DspCLI in module DSP_submission object:

class DspCLI(builtins.object)
 |  Methods defined here:
 |  
 |  __init__(self)
 |      Initialize self.  See help(type(self)) for accurate signature.
 |  
 |  create_new_team(self, description: str, centre_name: str) -> True
 |      Create a new team for the user.
 |      :param description : str
 |                           Brief description of the team
 |      :param centre_name : str
 |                           Name of the centre of the submission (e.g. EBI)
 |      :returns response : requests.Response
 |                          Response object from requests module
 |  
 |  create_submission(self, name: str = '') -> True
 |      Create an empty submission within the selected team.
 |      :param name: str
 |                   Name of the submission. If not specified the submission will be identified by its UUID.
 |      :returns response : requests.Response
 |                          Response object from requests module
 |  
 |  

But you can look at a specific question:

In [3]:
help(dsp.create_new_team)

Help on method create_new_team in module DSP_submission:

create_new_team(description: str, centre_name: str) -> True method of DSP_submission.DspCLI instance
    Create a new team for the user.
    :param description : str
                         Brief description of the team
    :param centre_name : str
                         Name of the centre of the submission (e.g. EBI)
    :returns response : requests.Response
                        Response object from requests module



## Creating and selecting a team and a submission

On this notebook, we're going to go through the whole process of a mock submission. In order to do so, we need to begin by creating a team:

In [4]:
dsp.create_new_team(description="Mock for notebook", centre_name="EBI")

<Response [201]>

Response 201 means that it has successfully been created. Now we will proceed to select the team:

In [5]:
dsp.select_team()

The teams are the following:
1 - subs.test-team-67
2 - subs.test-team-64
3 - subs.test-team-60
4 - subs.test-team-61
5 - subs.test-team-62
Please select a number: 1


{'name': 'subs.test-team-67',
 'description': 'Mock for notebook',
 'profile': {'centre name': 'EBI'},
 '_links': {'submissions': {'href': 'https://submission-test.ebi.ac.uk/api/submissions/search/by-team?teamName=subs.test-team-67'},
  'submissions:create': {'href': 'https://submission-test.ebi.ac.uk/api/teams/subs.test-team-67/submissions'},
  'items': {'href': 'https://submission-test.ebi.ac.uk/api/teams/subs.test-team-67/items'}}}

DSP assigns the name automatically, so it might be kinda hard to find your team if you have more than one. This function also returns a JSON with the team content, so this could be solved with a wrapper around this function.

Next step is to create a submission:

In [6]:
dsp.create_submission(name='Mock_submission')

<Response [201]>

And select it:

In [7]:
dsp.select_submission()

Submissions available for team 'subs.test-team-67' are the following:
1 - Name: Mock_submission
Please select a number: 1


{'id': 'bbab567e-ca24-4560-a683-7259e5a1b442',
 'submitter': {'email': 'enrique@ebi.ac.uk', 'name': 'Enrique Sapena Ventura'},
 'team': {'name': 'subs.test-team-67',
  'description': 'Mock for notebook',
  'profile': {'centre name': 'EBI'},
  '_links': {'submissions': {'href': 'https://submission-test.ebi.ac.uk/api/submissions/search/by-team?teamName=subs.test-team-67'},
   'submissions:create': {'href': 'https://submission-test.ebi.ac.uk/api/teams/subs.test-team-67/submissions'},
   'items': {'href': 'https://submission-test.ebi.ac.uk/api/teams/subs.test-team-67/items'}}},
 'createdDate': '2020-02-21T16:07:08.688+0000',
 'lastModifiedDate': '2020-02-21T16:07:08.688+0000',
 'createdBy': 'usr-19dace37-479e-4ce3-b661-3bc7901eb996',
 'lastModifiedBy': 'usr-19dace37-479e-4ce3-b661-3bc7901eb996',
 'name': 'Mock_submission',
 '_links': {'self': {'href': 'https://submission-test.ebi.ac.uk/api/submissions/bbab567e-ca24-4560-a683-7259e5a1b442'},
  'submission': {'href': 'https://submission-test

This returns the submission, but you don't need to worry about that. 

If, at some point, you just want to look at what submissions are available for a team, you can also run:

In [8]:
dsp.show_submissions()

Submissions available for team 'subs.test-team-67' are the following:
1 - Name: Mock_submission


[{'name': 'Mock_submission',
  'team': 'subs.test-team-67',
  'createdDate': '2020-02-21T16:07:08.688+0000',
  'lastModifiedDate': '2020-02-21T16:07:08.688+0000',
  'lastModifiedBy': 'usr-19dace37-479e-4ce3-b661-3bc7901eb996',
  'submissionStatus': 'Draft',
  'createdBy': 'usr-19dace37-479e-4ce3-b661-3bc7901eb996',
  'submitter': 'enrique@ebi.ac.uk',
  '_links': {'self': {'href': 'https://submission-test.ebi.ac.uk/api/submissions/bbab567e-ca24-4560-a683-7259e5a1b442'},
   'submission': {'href': 'https://submission-test.ebi.ac.uk/api/submissions/bbab567e-ca24-4560-a683-7259e5a1b442{?projection}',
    'templated': True},
   'submissionStatus': {'href': 'https://submission-test.ebi.ac.uk/api/submissions/bbab567e-ca24-4560-a683-7259e5a1b442/submissionStatus'},
   'submissionPlan': {'href': 'https://submission-test.ebi.ac.uk/api/submissions/bbab567e-ca24-4560-a683-7259e5a1b442/submissionPlan'}}}]

Or show the available teams for the user:

In [9]:
dsp.show_teams()

The teams are the following:
1 - subs.test-team-67
2 - subs.test-team-64
3 - subs.test-team-60
4 - subs.test-team-61
5 - subs.test-team-62


[{'name': 'subs.test-team-67',
  'description': 'Mock for notebook',
  'profile': {'centre name': 'EBI'},
  '_links': {'submissions': {'href': 'https://submission-test.ebi.ac.uk/api/submissions/search/by-team?teamName=subs.test-team-67'},
   'submissions:create': {'href': 'https://submission-test.ebi.ac.uk/api/teams/subs.test-team-67/submissions'},
   'items': {'href': 'https://submission-test.ebi.ac.uk/api/teams/subs.test-team-67/items'}}},
 {'name': 'subs.test-team-64',
  'description': 'New_team_to_test',
  'profile': {'centre name': 'Test to see if new team needed per submission'},
  '_links': {'submissions': {'href': 'https://submission-test.ebi.ac.uk/api/submissions/search/by-team?teamName=subs.test-team-64'},
   'submissions:create': {'href': 'https://submission-test.ebi.ac.uk/api/teams/subs.test-team-64/submissions'},
   'items': {'href': 'https://submission-test.ebi.ac.uk/api/teams/subs.test-team-64/items'}}},
 {'name': 'subs.test-team-60',
  'description': 'EBI-Test',
  'prof

# I have my submission created. Now what?

Once you have your submission created and you have selected it (Not necessary when creating), the next step is to determine where are your submittables. **This guide assumes you already have the submittable JSONs ready**.

As you might (Or might not) know, the DSP divides the "submittables" in 5 different categories, which will be validated differently. We have a hardcoded list of accepted submittables as an attribute of the object:

In [10]:
dsp.show_accepted_submittables()

1 - projects
2 - samples
3 - study
4 - assays
5 - assay_data


From here, you have 2 options:

1. Push the submittables from a directory with the function `self.submit_directory(directory_name)`
1. Push the submittables one by one with the function `self.create_submittable(json_content, submittable_type).

The first one is strongly discouraged as it requires all the submittables to have a filename like `<submittable_type>__<submittable_name>.json` (e.g. `samples__cell_suspension_1.json`) and doesn't account for validation errors due to sample linking.

For the purpose of this walkthrough, we will submit one by one by using the first function. As the json content, you can either pass a python dictionary with the content of the submittable or a string with the path of a JSON file. We will also check that the submittable has been created correctly:

In [ ]:
submittables_directory = '/Users/enrique/HumanCellAtlas/hca-to-dsp-tools/' # This is where the submittables are

submittables = ['assay_data__5386STDY7557335.bam.json',
                'assays__lib_5.json',
                'projects__EmbryonicHindlimb.json',
                'samples__BRC2091.json',
                'samples__C4-FSM-0-SC-1.json',
                'samples__cell_suspension_5.json',
                'study__EmbryonicHindlimb.json']
